// @ctrl/ngx-github-buttons/esm2022/ntkme/ntkme.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NtkmeCounterComponent } from './ntkme-counter.component';
import { NtkmeButtonComponent } from './ntkme.component';
import * as i0 from "@angular/core";
class NtkmeButtonModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: NtkmeButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.1", ngImport: i0, type: NtkmeButtonModule, declarations: [NtkmeButtonComponent, NtkmeCounterComponent], imports: [CommonModule], exports: [NtkmeButtonComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: NtkmeButtonModule, imports: [CommonModule] });
}
export { NtkmeButtonModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: NtkmeButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [NtkmeButtonComponent],
                    declarations: [NtkmeButtonComponent, NtkmeCounterComponent],
                }]
        }] });
