// @ctrl/ngx-github-buttons/esm2022/mdo/mdo.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { MdoCounterComponent } from './mdo-counter.component';
import { MdoGithubButtonComponent } from './mdo.component';
import * as i0 from "@angular/core";
class MdoButtonModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: MdoButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.1", ngImport: i0, type: MdoButtonModule, declarations: [MdoGithubButtonComponent, MdoCounterComponent], imports: [CommonModule], exports: [MdoGithubButtonComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: MdoButtonModule, imports: [CommonModule] });
}
export { MdoButtonModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: MdoButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [MdoGithubButtonComponent],
                    declarations: [MdoGithubButtonComponent, MdoCounterComponent],
                }]
        }] });
