// @sutton-signwriting/sgnw-components/dist/components/index.js
/*!
 * The Sutton SignWriting Web Components
 */
export { setAssetPath, setPlatformOptions } from '@stencil/core/internal/client';
import { a as appGlobalScript } from './global.js';
export { FswButton, defineCustomElement as defineCustomElementFswButton } from './fsw-button.js';
export { FswPalette, defineCustomElement as defineCustomElementFswPalette } from './fsw-palette.js';
export { FswPaletteSymbol, defineCustomElement as defineCustomElementFswPaletteSymbol } from './fsw-palette-symbol.js';
export { FswSign, defineCustomElement as defineCustomElementFswSign } from './fsw-sign.js';
export { FswSignbox, defineCustomElement as defineCustomElementFswSignbox } from './fsw-signbox.js';
export { FswSpatial, defineCustomElement as defineCustomElementFswSpatial } from './fsw-spatial.js';
export { FswSymbol, defineCustomElement as defineCustomElementFswSymbol } from './fsw-symbol.js';
export { FswVp, defineCustomElement as defineCustomElementFswVp } from './fsw-vp.js';
export { SgnwButton, defineCustomElement as defineCustomElementSgnwButton } from './sgnw-button.js';
export { SgnwPalette, defineCustomElement as defineCustomElementSgnwPalette } from './sgnw-palette.js';
export { SgnwPaletteSymbol, defineCustomElement as defineCustomElementSgnwPaletteSymbol } from './sgnw-palette-symbol.js';
export { SgnwSign, defineCustomElement as defineCustomElementSgnwSign } from './sgnw-sign.js';
export { SgnwSignbox, defineCustomElement as defineCustomElementSgnwSignbox } from './sgnw-signbox.js';
export { SgnwSymbol, defineCustomElement as defineCustomElementSgnwSymbol } from './sgnw-symbol.js';
export { SgnwVp, defineCustomElement as defineCustomElementSgnwVp } from './sgnw-vp.js';

const globalScripts = appGlobalScript;

globalScripts();
