// @ctrl/ngx-github-buttons/esm2022/gh/gh.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { GhCounterComponent } from './gh-counter.component';
import { GhButtonComponent } from './gh.component';
import * as i0 from "@angular/core";
class GhButtonModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: GhButtonModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "16.0.1", ngImport: i0, type: GhButtonModule, declarations: [GhButtonComponent, GhCounterComponent], imports: [CommonModule], exports: [GhButtonComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: GhButtonModule, imports: [CommonModule] });
}
export { GhButtonModule };
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.0.1", ngImport: i0, type: GhButtonModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    exports: [GhButtonComponent],
                    declarations: [GhButtonComponent, GhCounterComponent],
                }]
        }] });
